[ClangImporter] Hack. Break deserialization cycle with NSErrorPointer on iOS and watchOS.

huonw opened the pull request:
Addresses rdar://problem/32226279 .

I haven't yet worked out why this is only occurring on iOS and watchOS, and this is completely not the right long-term fix.

DougGregor commented:
This is gross :(

jrose-apple commented:
I think this'll change how we import `NSError **` when we can't turn it into `throws`, which means we need to come up with something a little better here.

jrose-apple commented:
Sorry, I skipped a step: good working tracking this down! I'm curious about which declaration is causing the cycle too, but that might not be something we can post on GitHub.

ematejska commented:
@swift-ci smoke test

huonw commented:
Wrote a comment on rdar://problem/32226279 ; move discussion there?

huonw committed:
[ClangImporter] Don't resugar NSPointer for the error out parameter of throws functions.

This breaks a cycle with a function in Foundation, but isn't a complete
fix: it will be similarly problematic if a function that can't have the
throws conversion is added.

huonw commented:
@swift-ci please smoke test

milseman commented:
@huonw please file an SR for the less hacky solution

milseman commented:
@jrose-apple should this be nominated for the 4.0 branch?

jrose-apple commented:
Yes, I think so.

milseman commented:
I will nominate this

jrose-apple commented:
Thanks, Michael!

milseman commented:
@jrose-apple Could you or @huonw help me fill in the explanation, scope, and risk?
